File: Engine/src/tungsten/wolfram_processes.py
from __future__ import annotations

import json
import os
import subprocess
import tempfile
import time
from contextlib import contextmanager
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterator

try:
    import msvcrt
except ImportError:  # pragma: no cover - non-Windows fallback
    msvcrt = None  # type: ignore[assignment]


_WOLFRAM_PROCESS_NAMES = {
    "mathkernel",
    "mathkernel.exe",
    "mathematica",
    "mathematica.exe",
    "wolfram",
    "wolfram.exe",
    "wolframdesktop",
    "wolframdesktop.exe",
    "wolframkernel",
    "wolframkernel.exe",
}


@dataclass(frozen=True)
class WolframProcessInfo:
    pid: int
    parent_pid: int
    name: str
    executable_path: str | None
    command_line: str | None
    started_utc: str | None
    tungsten_owned: bool
    headless_batch: bool
    parent_missing: bool
    controlling_process_candidate: bool

    @property
    def age_seconds(self) -> float | None:
        if self.started_utc is None:
            return None
        try:
            started = datetime.fromisoformat(self.started_utc.replace("Z", "+00:00"))
        except ValueError:
            return None
        return max(0.0, (datetime.now(timezone.utc) - started).total_seconds())

    def to_dict(self) -> dict[str, object]:
        return {
            "pid": self.pid,
            "parent_pid": self.parent_pid,
            "name": self.name,
            "executable_path": self.executable_path,
            "command_line": self.command_line,
            "started_utc": self.started_utc,
            "tungsten_owned": self.tungsten_owned,
            "headless_batch": self.headless_batch,
            "parent_missing": self.parent_missing,
            "controlling_process_candidate": self.controlling_process_candidate,
            "age_seconds": self.age_seconds,
        }


@dataclass(frozen=True)
class WolframProcessSnapshot:
    processes: tuple[WolframProcessInfo, ...]
    cached_max_license_processes: int | None

    @property
    def active_count(self) -> int:
        return sum(1 for process in self.processes if process.controlling_process_candidate)

    def to_dict(self) -> dict[str, object]:
        return {
            "cached_max_license_processes": self.cached_max_license_processes,
            "active_count": self.active_count,
            "processes": [process.to_dict() for process in self.processes],
        }


def _cache_root() -> Path:
    local_app_data = os.environ.get("LOCALAPPDATA")
    if local_app_data:
        return Path(local_app_data) / "Tungsten"
    return Path(tempfile.gettempdir()) / "Tungsten"


def _cache_path() -> Path:
    return _cache_root() / "wolfram-license-cache.json"


def read_cached_max_license_processes() -> int | None:
    path = _cache_path()
    if not path.exists():
        return None
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None
    value = payload.get("max_license_processes")
    if isinstance(value, int) and value > 0:
        return value
    return None


def write_cached_max_license_processes(value: int) -> None:
    if value <= 0:
        return
    path = _cache_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "max_license_processes": value,
        "updated_utc": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
    }
    path.write_text(json.dumps(payload, indent=2), encoding="utf-8")


def _powershell_executable() -> str:
    for candidate in ("pwsh", "powershell"):
        try:
            completed = subprocess.run(
                [candidate, "-NoLogo", "-NoProfile", "-Command", "$PSVersionTable.PSVersion.ToString()"],
                capture_output=True,
                text=True,
                encoding="utf-8",
                errors="replace",
                timeout=5,
            )
        except (OSError, subprocess.TimeoutExpired):
            continue
        if completed.returncode == 0:
            return candidate
    raise RuntimeError("Could not find PowerShell for Wolfram process inspection.")


def _normalize_process_payload(payload: object) -> list[dict[str, object]]:
    if payload is None:
        return []
    if isinstance(payload, list):
        return [item for item in payload if isinstance(item, dict)]
    if isinstance(payload, dict):
        return [payload]
    return []


def _is_wolfram_process(name: str | None, executable_path: str | None) -> bool:
    if name and name.lower() in _WOLFRAM_PROCESS_NAMES:
        return True
    if executable_path and "wolfram" in executable_path.lower():
        return True
    return False


def _is_controlling_process_candidate(name: str, lower_command: str) -> bool:
    normalized = name.lower()
    if normalized in {"mathematica", "mathematica.exe", "wolframdesktop", "wolframdesktop.exe"}:
        return True
    if normalized in {"wolfram", "wolfram.exe"}:
        return not any(marker in lower_command for marker in (" -mathlink ", " -subkernel ", "playerpass"))
    if normalized in {"wolframkernel", "wolframkernel.exe", "mathkernel", "mathkernel.exe"}:
        return not any(marker in lower_command for marker in (" -mathlink ", " -subkernel ", "playerpass"))
    return False


def list_wolfram_processes() -> list[WolframProcessInfo]:
    script = r"""
$all = Get-CimInstance Win32_Process | Select-Object `
    Name, ProcessId, ParentProcessId, ExecutablePath, CommandLine,
    @{ Name = "StartedUtc"; Expression = {
        if ($_.CreationDate) {
            [System.Management.ManagementDateTimeConverter]::ToDateTime($_.CreationDate).ToUniversalTime().ToString("o")
        }
        else {
            $null
        }
    }}
$all | ConvertTo-Json -Compress -Depth 3
"""
    completed = subprocess.run(
        [_powershell_executable(), "-NoLogo", "-NoProfile", "-Command", script],
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
        timeout=15,
    )
    if completed.returncode != 0:
        raise RuntimeError(f"PowerShell process inspection failed: {completed.stderr or completed.stdout}")

    raw = completed.stdout.strip()
    payload = json.loads(raw) if raw else []
    rows = _normalize_process_payload(payload)
    live_pids = {
        int(row["ProcessId"])
        for row in rows
        if isinstance(row.get("ProcessId"), int)
    }

    processes: list[WolframProcessInfo] = []
    for row in rows:
        name = row.get("Name")
        executable_path = row.get("ExecutablePath")
        if not _is_wolfram_process(str(name) if name is not None else None, str(executable_path) if executable_path is not None else None):
            continue

        command_line = str(row["CommandLine"]) if row.get("CommandLine") is not None else None
        lower_command = (command_line or "").lower()
        tungsten_owned = any(marker in lower_command for marker in ("tungsten-wrapper-", "tungsten-mathpass-"))
        headless_batch = any(marker in lower_command for marker in (" -script ", " -run ", " -runfile "))
        parent_pid = int(row["ParentProcessId"]) if isinstance(row.get("ParentProcessId"), int) else 0
        parent_missing = parent_pid > 0 and parent_pid not in live_pids
        controlling_process_candidate = _is_controlling_process_candidate(str(name), lower_command)

        processes.append(
            WolframProcessInfo(
                pid=int(row["ProcessId"]),
                parent_pid=parent_pid,
                name=str(name),
                executable_path=str(executable_path) if executable_path is not None else None,
                command_line=command_line,
                started_utc=str(row["StartedUtc"]) if row.get("StartedUtc") is not None else None,
                tungsten_owned=tungsten_owned,
                headless_batch=headless_batch,
                parent_missing=parent_missing,
                controlling_process_candidate=controlling_process_candidate,
            )
        )

    return sorted(processes, key=lambda process: process.pid)


def snapshot_wolfram_processes() -> WolframProcessSnapshot:
    return WolframProcessSnapshot(
        processes=tuple(list_wolfram_processes()),
        cached_max_license_processes=read_cached_max_license_processes(),
    )


def cleanup_stale_tungsten_processes(min_age_seconds: float = 30.0) -> list[int]:
    cleaned: list[int] = []
    for process in list_wolfram_processes():
        age = process.age_seconds
        if not process.tungsten_owned:
            continue
        if not process.headless_batch:
            continue
        if not process.parent_missing:
            continue
        if age is not None and age < min_age_seconds:
            continue

        completed = subprocess.run(
            ["taskkill", "/PID", str(process.pid), "/T", "/F"],
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=15,
        )
        if completed.returncode == 0:
            cleaned.append(process.pid)
    return cleaned


@contextmanager
def tungsten_wolfram_launch_gate(timeout_seconds: float = 900.0, poll_seconds: float = 0.2) -> Iterator[float]:
    if msvcrt is None:  # pragma: no cover - non-Windows fallback
        yield 0.0
        return

    lock_path = _cache_root() / "wolfram-launch.lock"
    lock_path.parent.mkdir(parents=True, exist_ok=True)
    waited = 0.0
    start = time.monotonic()

    with lock_path.open("a+b") as handle:
        handle.seek(0, os.SEEK_END)
        if handle.tell() == 0:
            handle.write(b"0")
            handle.flush()
        while True:
            try:
                handle.seek(0)
                msvcrt.locking(handle.fileno(), msvcrt.LK_NBLCK, 1)
                break
            except OSError:
                waited = time.monotonic() - start
                if waited >= timeout_seconds:
                    raise TimeoutError("Timed out waiting for the Tungsten Wolfram launch gate.")
                time.sleep(poll_seconds)

        try:
            waited = time.monotonic() - start
            yield waited
        finally:
            handle.seek(0)
            msvcrt.locking(handle.fileno(), msvcrt.LK_UNLCK, 1)


def wait_for_wolfram_license_slot(
    cached_max_license_processes: int | None,
    *,
    timeout_seconds: float = 30.0,
    poll_seconds: float = 0.5,
) -> tuple[WolframProcessSnapshot, float, bool]:
    start = time.monotonic()
    snapshot = snapshot_wolfram_processes()
    if cached_max_license_processes is None or snapshot.active_count < cached_max_license_processes:
        return snapshot, 0.0, True

    while True:
        elapsed = time.monotonic() - start
        if elapsed >= timeout_seconds:
            return snapshot, elapsed, False
        time.sleep(poll_seconds)
        snapshot = snapshot_wolfram_processes()
        if snapshot.active_count < cached_max_license_processes:
            return snapshot, time.monotonic() - start, True

File: Engine/haskell/src/Tungsten/WolframProcesses.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Wolfram process inspection, cached license limits, and launch coordination.
module Tungsten.WolframProcesses
  ( WolframProcessInfo (..)
  , WolframProcessSnapshot (..)
  , wolframProcessAgeSeconds
  , wolframProcessPayload
  , wolframProcessSnapshotPayload
  , activeWolframProcessCount
  , isControllingProcessCandidate
  , readCachedMaxLicenseProcesses
  , readCachedMaxLicenseProcessesAt
  , writeCachedMaxLicenseProcesses
  , writeCachedMaxLicenseProcessesAt
  , listWolframProcesses
  , snapshotWolframProcesses
  , cleanupStaleTungstenProcesses
  , withWolframLaunchGate
  , waitForWolframLicenseSlot
  , waitForWolframLicenseSlotWith
  ) where

import Control.Concurrent (threadDelay)
import Control.Exception (IOException, bracket, try)
import Control.Monad (forM)
import qualified Data.Map.Strict as Map
import Data.Maybe (catMaybes)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TextIO
import Data.Time.Clock (UTCTime, diffUTCTime, getCurrentTime)
import Data.Time.Format (defaultTimeLocale, formatTime, parseTimeM)
import System.Directory
  ( createDirectory
  , createDirectoryIfMissing
  , doesDirectoryExist
  , doesFileExist
  , findExecutable
  , getTemporaryDirectory
  , removeDirectory
  )
import System.Environment (lookupEnv)
import System.Exit (ExitCode (..))
import System.FilePath ((</>), takeDirectory)
import System.Info (os)
import System.Process (readProcessWithExitCode)
import Text.Read (readMaybe)
import Tungsten.Json

data WolframProcessInfo = WolframProcessInfo
  { wolframProcessPid :: !Int
  , wolframProcessParentPid :: !Int
  , wolframProcessName :: !Text
  , wolframProcessExecutablePath :: !(Maybe FilePath)
  , wolframProcessCommandLine :: !(Maybe Text)
  , wolframProcessStartedUtc :: !(Maybe Text)
  , wolframProcessTungstenOwned :: !Bool
  , wolframProcessHeadlessBatch :: !Bool
  , wolframProcessParentMissing :: !Bool
  , wolframProcessControllingCandidate :: !Bool
  }
  deriving (Eq, Show)

data WolframProcessSnapshot = WolframProcessSnapshot
  { wolframSnapshotProcesses :: ![WolframProcessInfo]
  , wolframSnapshotCachedMaxLicenseProcesses :: !(Maybe Int)
  }
  deriving (Eq, Show)

wolframProcessAgeSeconds :: UTCTime -> WolframProcessInfo -> Maybe Double
wolframProcessAgeSeconds now process = do
  source <- wolframProcessStartedUtc process
  started <-
    parseTimeM True defaultTimeLocale "%Y-%m-%dT%H:%M:%S%QZ" (T.unpack source)
      :: Maybe UTCTime
  pure (max 0 (realToFrac (diffUTCTime now started)))

wolframProcessPayload :: UTCTime -> WolframProcessInfo -> JsonValue
wolframProcessPayload now process =
  JsonObject
    ( Map.fromList
        [ ("age_seconds", maybe JsonNull jsonDouble (wolframProcessAgeSeconds now process))
        , ("command_line", maybe JsonNull JsonString (wolframProcessCommandLine process))
        , ("controlling_process_candidate", JsonBool (wolframProcessControllingCandidate process))
        , ("executable_path", maybe JsonNull (JsonString . T.pack) (wolframProcessExecutablePath process))
        , ("headless_batch", JsonBool (wolframProcessHeadlessBatch process))
        , ("name", JsonString (wolframProcessName process))
        , ("parent_missing", JsonBool (wolframProcessParentMissing process))
        , ("parent_pid", jsonInteger (fromIntegral (wolframProcessParentPid process)))
        , ("pid", jsonInteger (fromIntegral (wolframProcessPid process)))
        , ("started_utc", maybe JsonNull JsonString (wolframProcessStartedUtc process))
        , ("tungsten_owned", JsonBool (wolframProcessTungstenOwned process))
        ]
    )

wolframProcessSnapshotPayload :: UTCTime -> WolframProcessSnapshot -> JsonValue
wolframProcessSnapshotPayload now snapshot =
  JsonObject
    ( Map.fromList
        [ ("active_count", jsonInteger (fromIntegral (activeWolframProcessCount snapshot)))
        , ( "cached_max_license_processes"
          , maybe JsonNull (jsonInteger . fromIntegral) (wolframSnapshotCachedMaxLicenseProcesses snapshot)
          )
        , ("processes", JsonArray (map (wolframProcessPayload now) (wolframSnapshotProcesses snapshot)))
        ]
    )

activeWolframProcessCount :: WolframProcessSnapshot -> Int
activeWolframProcessCount =
  length . filter wolframProcessControllingCandidate . wolframSnapshotProcesses

isControllingProcessCandidate :: Text -> Text -> Bool
isControllingProcessCandidate name commandLine
  | normalized `elem` ["mathematica", "mathematica.exe", "wolframdesktop", "wolframdesktop.exe"] = True
  | normalized `elem` ["wolfram", "wolfram.exe", "wolframkernel", "wolframkernel.exe", "mathkernel", "mathkernel.exe"] =
      not (any (`T.isInfixOf` lowerCommand) [" -mathlink ", " -subkernel ", "playerpass"])
  | otherwise = False
 where
  normalized = T.toLower name
  lowerCommand = T.toLower commandLine

readCachedMaxLicenseProcesses :: IO (Maybe Int)
readCachedMaxLicenseProcesses = cachePath >>= readCachedMaxLicenseProcessesAt

readCachedMaxLicenseProcessesAt :: FilePath -> IO (Maybe Int)
readCachedMaxLicenseProcessesAt path = do
  exists <- doesFileExist path
  if not exists
    then pure Nothing
    else do
      sourceResult <- try (TextIO.readFile path)
      pure $ case (sourceResult :: Either IOException Text) of
        Left _ -> Nothing
        Right source -> case parseJson source of
          Right (JsonObject payload) -> case Map.lookup "max_license_processes" payload of
            Just (JsonNumber value) -> readMaybe (T.unpack value) >>= positive
            _ -> Nothing
          _ -> Nothing
 where
  positive value = if value > 0 then Just value else Nothing

writeCachedMaxLicenseProcesses :: Int -> IO ()
writeCachedMaxLicenseProcesses value = cachePath >>= \path -> writeCachedMaxLicenseProcessesAt path value

writeCachedMaxLicenseProcessesAt :: FilePath -> Int -> IO ()
writeCachedMaxLicenseProcessesAt _ value | value <= 0 = pure ()
writeCachedMaxLicenseProcessesAt path value = do
  createDirectoryIfMissing True (takeDirectory path)
  now <- getCurrentTime
  let payload =
        JsonObject
          ( Map.fromList
              [ ("max_license_processes", jsonInteger (fromIntegral value))
              , ("updated_utc", JsonString (T.pack (formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" now)))
              ]
          )
  TextIO.writeFile path (encodeJson payload <> "\n")

listWolframProcesses :: IO [WolframProcessInfo]
listWolframProcesses
  | os /= "mingw32" = pure []
  | otherwise = do
      executable <- findPowerShell
      case executable of
        Nothing -> pure []
        Just command -> do
          completed <- try (readProcessWithExitCode command powershellArguments "")
          pure $ case (completed :: Either IOException (ExitCode, String, String)) of
            Right (ExitSuccess, stdoutText, _) -> decodeProcessRows (T.pack stdoutText)
            _ -> []

snapshotWolframProcesses :: IO WolframProcessSnapshot
snapshotWolframProcesses =
  WolframProcessSnapshot <$> listWolframProcesses <*> readCachedMaxLicenseProcesses

cleanupStaleTungstenProcesses :: Double -> IO [Int]
cleanupStaleTungstenProcesses minimumAge = do
  now <- getCurrentTime
  processes <- listWolframProcesses
  fmap catMaybes . forM processes $ \process ->
    if stale now process
      then terminateProcessTree (wolframProcessPid process)
      else pure Nothing
 where
  stale now process =
    wolframProcessTungstenOwned process
      && wolframProcessHeadlessBatch process
      && wolframProcessParentMissing process
      && maybe True (>= minimumAge) (wolframProcessAgeSeconds now process)

withWolframLaunchGate
  :: Double
  -> Double
  -> (Double -> IO value)
  -> IO (Either Text value)
withWolframLaunchGate timeoutSeconds pollSeconds action = do
  root <- cacheRoot
  let lockDirectory = root </> "wolfram-launch.lock"
  createDirectoryIfMissing True root
  started <- getCurrentTime
  acquire lockDirectory started
 where
  acquire lockDirectory started = do
    acquired <- try (createDirectory lockDirectory)
    case (acquired :: Either IOException ()) of
      Right () ->
        Right <$> bracket (pure lockDirectory) release (const (elapsedSince started >>= action))
      Left _ -> do
        now <- getCurrentTime
        let waited = realToFrac (diffUTCTime now started)
        if waited >= timeoutSeconds
          then pure (Left "Timed out waiting for the Tungsten Wolfram launch gate.")
          else threadDelay (secondsToMicroseconds pollSeconds) >> acquire lockDirectory started
  release lockDirectory = do
    exists <- doesDirectoryExist lockDirectory
    if exists then removeDirectory lockDirectory else pure ()

waitForWolframLicenseSlot
  :: Maybe Int
  -> Double
  -> Double
  -> IO (WolframProcessSnapshot, Double, Bool)
waitForWolframLicenseSlot = waitForWolframLicenseSlotWith snapshotWolframProcesses

waitForWolframLicenseSlotWith
  :: IO WolframProcessSnapshot
  -> Maybe Int
  -> Double
  -> Double
  -> IO (WolframProcessSnapshot, Double, Bool)
waitForWolframLicenseSlotWith snapshotAction cachedMaximum timeoutSeconds pollSeconds = do
  started <- getCurrentTime
  firstSnapshot <- snapshotAction
  case cachedMaximum of
    Nothing -> pure (firstSnapshot, 0, True)
    Just maximumProcesses
      | activeWolframProcessCount firstSnapshot < maximumProcesses -> pure (firstSnapshot, 0, True)
      | otherwise -> poll started maximumProcesses firstSnapshot
 where
  poll started maximumProcesses snapshot = do
    waited <- elapsedSince started
    if waited >= timeoutSeconds
      then pure (snapshot, waited, False)
      else do
        threadDelay (secondsToMicroseconds pollSeconds)
        next <- snapshotAction
        if activeWolframProcessCount next < maximumProcesses
          then do
            finishedWait <- elapsedSince started
            pure (next, finishedWait, True)
          else poll started maximumProcesses next

cacheRoot :: IO FilePath
cacheRoot = do
  localAppData <- lookupEnv "LOCALAPPDATA"
  case localAppData of
    Just path | not (null path) -> pure (path </> "Tungsten")
    _ -> (</> "Tungsten") <$> getTemporaryDirectory

cachePath :: IO FilePath
cachePath = (</> "wolfram-license-cache.json") <$> cacheRoot

findPowerShell :: IO (Maybe FilePath)
findPowerShell = findExecutable "pwsh" >>= \case
  Just path -> pure (Just path)
  Nothing -> findExecutable "powershell"

powershellArguments :: [String]
powershellArguments =
  [ "-NoLogo", "-NoProfile", "-Command"
  , "$all = Get-CimInstance Win32_Process | Select-Object Name, ProcessId, ParentProcessId, ExecutablePath, CommandLine, @{Name='StartedUtc';Expression={if ($_.CreationDate) {[System.Management.ManagementDateTimeConverter]::ToDateTime($_.CreationDate).ToUniversalTime().ToString('o')} else {$null}}}; $all | ConvertTo-Json -Compress -Depth 3"
  ]

decodeProcessRows :: Text -> [WolframProcessInfo]
decodeProcessRows source = case parseJson (T.strip source) of
  Right (JsonArray rows) -> buildRows rows
  Right row@JsonObject {} -> buildRows [row]
  _ -> []
 where
  buildRows rows =
    let livePids = [pid | JsonObject row <- rows, Just pid <- [integerField "ProcessId" row]]
     in catMaybes (map (decodeProcess livePids) rows)

decodeProcess :: [Int] -> JsonValue -> Maybe WolframProcessInfo
decodeProcess livePids (JsonObject row) = do
  pid <- integerField "ProcessId" row
  name <- stringField "Name" row
  let executablePath = T.unpack <$> optionalStringField "ExecutablePath" row
      commandLine = optionalStringField "CommandLine" row
      lowerCommand = T.toLower (maybe "" id commandLine)
      parentPid = maybe 0 id (integerField "ParentProcessId" row)
      isWolfram = T.toLower name `elem` wolframNames || maybe False (T.isInfixOf "wolfram" . T.toLower . T.pack) executablePath
  if not isWolfram
    then Nothing
    else
      Just
        WolframProcessInfo
          { wolframProcessPid = pid
          , wolframProcessParentPid = parentPid
          , wolframProcessName = name
          , wolframProcessExecutablePath = executablePath
          , wolframProcessCommandLine = commandLine
          , wolframProcessStartedUtc = optionalStringField "StartedUtc" row
          , wolframProcessTungstenOwned = any (`T.isInfixOf` lowerCommand) ["tungsten-wrapper-", "tungsten-mathpass-"]
          , wolframProcessHeadlessBatch = any (`T.isInfixOf` lowerCommand) [" -script ", " -run ", " -runfile "]
          , wolframProcessParentMissing = parentPid > 0 && parentPid `notElem` livePids
          , wolframProcessControllingCandidate = isControllingProcessCandidate name lowerCommand
          }
decodeProcess _ _ = Nothing

integerField :: Text -> Map.Map Text JsonValue -> Maybe Int
integerField key payload = case Map.lookup key payload of
  Just (JsonNumber value) -> readMaybe (T.unpack value)
  _ -> Nothing

stringField :: Text -> Map.Map Text JsonValue -> Maybe Text
stringField key payload = case Map.lookup key payload of
  Just (JsonString value) -> Just value
  _ -> Nothing

optionalStringField :: Text -> Map.Map Text JsonValue -> Maybe Text
optionalStringField = stringField

terminateProcessTree :: Int -> IO (Maybe Int)
terminateProcessTree pid
  | os /= "mingw32" = pure Nothing
  | otherwise = do
      completed <- try (readProcessWithExitCode "taskkill" ["/PID", show pid, "/T", "/F"] "")
      pure $ case (completed :: Either IOException (ExitCode, String, String)) of
        Right (ExitSuccess, _, _) -> Just pid
        _ -> Nothing

elapsedSince :: UTCTime -> IO Double
elapsedSince started = do
  now <- getCurrentTime
  pure (realToFrac (diffUTCTime now started))

secondsToMicroseconds :: Double -> Int
secondsToMicroseconds seconds = max 0 (floor (seconds * 1000000))

wolframNames :: [Text]
wolframNames =
  [ "mathkernel", "mathkernel.exe", "mathematica", "mathematica.exe"
  , "wolfram", "wolfram.exe", "wolframdesktop", "wolframdesktop.exe"
  , "wolframkernel", "wolframkernel.exe"
  ]

jsonInteger :: Integer -> JsonValue
jsonInteger = JsonNumber . T.pack . show

jsonDouble :: Double -> JsonValue
jsonDouble = JsonNumber . T.pack . show
